### DISCUSSION TITLE
Question about the completeness of JAX operators.

### DISCUSSION THREAD

**Discussion by Dong-Jiahuan** [jax-ml/jax#18147] [Q&A] [answered]
I want to use JAX to do a project about AI. So, I'm curious about the completeness of JAX operators.
For now, is there any obvious shortcoming in the completeness of JAX operators compared to PyTorch or TensorFlow? 
For example, does there exist some models in AI that PyTorch or TensorFlow can build with their native operators but JAX can not and JAX users can only define custom operators using C++/CUDA?

(I know the positioning of JAX is different with PyTorch or TensorFlow. If an operator which is native in PyTorch or Tensorflow and it can be constructed with JAX native operators, I won't think the lack of that operator reflects the shortcoming in the completeness of JAX operators)

**Comment by pschuh** [comment, marked as answer]
Jax can lower to everything in XLA. Tensorflow also lowers a subset of its operations to XLA: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/tf2xla . Note that XLA is the only way to target TPUs using tensorflow, so for TPUs JAX and Tensorflow have full feature parity.  The TF operations that are not supported are all CPU/GPU-only operations like decoding jpegs, datasets, etc. XLA has one major limitation vs other approaches in that its shapes are all static (this helps XLA better plan tiling and buffer assignment). However, most ML models do not use dynamic shapes, and there are good workarounds like padding and masking for a limited 'bounded dynamism' regime.

**Reply by Dong-Jiahuan** [reply]
I see. Thanks for your reply!
